### PR TITLE
Add an example for connecting to a private SSM instance using PSC

### PR DESCRIPTION
**Pull request by ericpang777**
**Release Note Template for Downstream PRs (will be copied)**

```release-note:none
add an example for connecting to a private SSM instance using PSC
```

### CODE REVIEW

File: mmv1/products/securesourcemanager/Instance.yaml [hunk 1]
```diff
@@ -85,6 +85,32 @@ examples:
       instance_id: 'my-instance'
       ca_pool_id: 'ca-pool'
       root_ca_id: 'root-ca'
+  - !ruby/object:Provider::Terraform::Examples
+    name: 'secure_source_manager_instance_private_psc'
```
Comment (honghanmyb) [inline mmv1/products/securesourcemanager/Instance.yaml:89]: Let's split the examples. One for connecting using endpoint; and one for connecting using NEG and ILB.

File: mmv1/templates/terraform/examples/secure_source_manager_instance_private_psc.tf.erb [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+
+resource "google_compute_service_attachment" "service_attachment" {
+  name = "<%= ctx[:vars]['service_attachment_id'] %>"
+  region = "us-central1"
+  connection_preference = "ACCEPT_AUTOMATIC"
+  target_service = google_compute_forwarding_rule.fw_rule_target_proxy.id
+  nat_subnets = [google_compute_subnetwork.psc_subnet.id]
+  enable_proxy_protocol = false
+}
+
```
Comment (honghanmyb) [inline mmv1/templates/terraform/examples/secure_source_manager_instance_private_psc.tf.erb:86]: The extra psc setup part is not needed.

File: mmv1/templates/terraform/examples/secure_source_manager_instance_private_psc.tf.erb [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+resource "google_dns_record_set" "ssm_instance_git_record" {
+  name = "${google_secure_source_manager_instance.<%= ctx[:primary_resource_id] %>.host_config.0.git_http}."
+  type = "A"
+  ttl = 300
+  managed_zone = google_dns_managed_zone.private_zone.name
+  rrdatas = [google_compute_forwarding_rule.fw_rule_service_attachment.ip_address]
+}
+
+// Follow https://cloud.google.com/secure-source-manager/docs/create-private-service-connect-instance#root-ca-api
+resource "google_compute_instance" "compute_instance" {
```
Comment (honghanmyb) [inline mmv1/templates/terraform/examples/secure_source_manager_instance_private_psc.tf.erb:155]: I don't think we need this

File: mmv1/templates/terraform/examples/secure_source_manager_instance_private_psc.tf.erb [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+  scratch_disk {
+    interface = "NVME"
+  }
+  network_interface {
+    network = google_compute_network.network2.id
+    subnetwork = google_compute_subnetwork.subnet2.id
+  }
+}
+
+resource "google_privateca_ca_pool" "ca_pool" {
```
Comment (honghanmyb) [inline mmv1/templates/terraform/examples/secure_source_manager_instance_private_psc.tf.erb:174]: ca_pool and related resource should happen before creating ssm instance.

Comment (ericpang777) [reply mmv1/templates/terraform/examples/secure_source_manager_instance_private_psc.tf.erb]: Isn't this reflected in the dependency of the ssm instance, and it using the ca_pool?

File: mmv1/templates/terraform/examples/secure_source_manager_instance_private_psc.tf.erb (hunk 3 above)
Comment (ericpang777) [reply mmv1/templates/terraform/examples/secure_source_manager_instance_private_psc.tf.erb]: Removed

File: mmv1/products/securesourcemanager/Instance.yaml (hunk 1 above)
Comment (ericpang777) [reply mmv1/products/securesourcemanager/Instance.yaml]: Done, if you can double check

File: mmv1/products/securesourcemanager/Instance.yaml [hunk 5]
```diff
@@ -85,6 +85,41 @@ examples:
       instance_id: 'my-instance'
       ca_pool_id: 'ca-pool'
       root_ca_id: 'root-ca'
+  - !ruby/object:Provider::Terraform::Examples
+    name: 'secure_source_manager_instance_private_psc_backend'
+    external_providers: ["random", "time"]
```
Comment (roaks3) [inline mmv1/products/securesourcemanager/Instance.yaml:90]: Nit: I know this was probably copied from the others, but you aren't using the `random` provider (from what I can tell), so it should probably be removed.

**Review by roaks3 — APPROVED** [review]
I'm less familiar with the specifics of these resources, so I'll wait for @honghanmyb before merging anything, but LGTM on code and tests passing

File: mmv1/products/securesourcemanager/Instance.yaml (hunk 5 above)
Comment (ericpang777) [reply mmv1/products/securesourcemanager/Instance.yaml]: Removed. The previous was added by myself actually, although I can't remember why I added it...

**Review by roaks3 — APPROVED** [review]

File: mmv1/templates/terraform/examples/secure_source_manager_instance_private_psc_backend.tf.erb [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+  name = "<%= ctx[:vars]['network_id'] %>"
+  auto_create_subnetworks = false
+}
+
+resource "google_compute_subnetwork" "subnet" {
+  name = "<%= ctx[:vars]['subnet_id'] %>"
+  region = "us-central1"
+  network = google_compute_network.network.id
+  ip_cidr_range = "10.0.1.0/24"
+  private_ip_google_access = true # needed for "gcloud auth print-access-token" to work
```
Comment (honghanmyb) [inline mmv1/templates/terraform/examples/secure_source_manager_instance_private_psc_backend.tf.erb:12]: Remove the comment

File: mmv1/templates/terraform/examples/secure_source_manager_instance_private_psc_backend.tf.erb [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+}
+
+# ca pool IAM permissions can take time to propagate
+resource "time_sleep" "wait_120_seconds" {
+  depends_on = [google_privateca_ca_pool_iam_binding.ca_pool_binding]
+
+  create_duration = "120s"
+}
+
+data "google_project" "project" {}
```
Comment (honghanmyb) [inline mmv1/templates/terraform/examples/secure_source_manager_instance_private_psc_backend.tf.erb:178]: Move these to the top of the example. (project, cas resources, ssm instance etc..)

File: mmv1/templates/terraform/examples/secure_source_manager_instance_private_psc_endpoint.tf.erb [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+}
+
+# ca pool IAM permissions can take time to propagate
+resource "time_sleep" "wait_120_seconds" {
+  depends_on = [google_privateca_ca_pool_iam_binding.ca_pool_binding]
+
+  create_duration = "120s"
+}
+
+data "google_project" "project" {}
```
Comment (honghanmyb) [inline mmv1/templates/terraform/examples/secure_source_manager_instance_private_psc_endpoint.tf.erb:145]: Similarly, move these to the top of the example.

**Review by honghanmyb — APPROVED** [review]

File: mmv1/templates/terraform/examples/secure_source_manager_instance_private_psc_backend.tf.erb (hunk 7 above)
Comment (ericpang777) [reply mmv1/templates/terraform/examples/secure_source_manager_instance_private_psc_backend.tf.erb]: Done

File: mmv1/templates/terraform/examples/secure_source_manager_instance_private_psc_endpoint.tf.erb (hunk 8 above)
Comment (ericpang777) [reply mmv1/templates/terraform/examples/secure_source_manager_instance_private_psc_endpoint.tf.erb]: Done

File: mmv1/templates/terraform/examples/secure_source_manager_instance_private_psc_backend.tf.erb (hunk 6 above)
Comment (ericpang777) [reply mmv1/templates/terraform/examples/secure_source_manager_instance_private_psc_backend.tf.erb]: Done

**Comment by ericpang777** [comment]
> I'm less familiar with the specifics of these resources, so I'll wait for @honghanmyb before merging anything, but LGTM on code and tests passing

Han has approved now

**Pull request merged**